### PR TITLE
start loom dataset 

### LINKED ISSUE #105: Loom for RETINA dataset
Create a new class called `LoomDataset` that uses the loompy library to load an arbitrary dataset in the loom format.

https://github.com/linnarsson-lab/loompy

As a test case --- and so we can easily use it --- convert the RETINA dataset to loom format and upload it to the YosefLab/scVI-data repository.

### PR DESCRIPTION
**Pull request by imyiningliu**
I create test file using the retina test files in `test/data/` folder. I also create retina.loom using the raw data, which isn't uploaded here. 

closes #105

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -13,3 +13,4 @@ scipy==1.0.0
 h5py
 codecov
 pandas==0.23.0
+loompy==1.0.2
```
Comment (jeff-regier) [inline requirements.txt]: Let's require loompy 2.0.9

File: scvi/dataset/__init__.py [hunk 2]
```diff
@@ -36,6 +38,8 @@ def load_datasets(dataset_name, unit_test=False):
         gene_dataset = HematoDataset(unit_test=unit_test)
     elif dataset_name == 'pbmc':
         gene_dataset = PbmcDataset(unit_test=unit_test)
+    elif dataset_name == 'loom':
```
Comment (jeff-regier) [inline scvi/dataset/__init__.py]: How about passing the loom file name here (e.g., `/home/jeff/data/retina.loom`) as the dataset set name, and then branching based on the suffix? e.g.
```
elif dataset_name[-5:] == ".loom":
    gene_dataset = LoomDataset(dataset_name, unit_test=unit_test)
```

File: scvi/dataset/loom.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        self.save_path = 'data/' if not unit_test else 'tests/data/'
+
+        data, labels, cell_batches = self.preprocess()
+        if unit_test:
+            cell_batches = np.reshape(cell_batches, (5, 1))
+            cell_batches[0] = 1
+            data = np.reshape(data, (5, 13166))
+            data = data.repeat(100, 0)
+            cell_batches = cell_batches.repeat(100, 0)
+            labels = np.arange(5).reshape(5, 1).repeat(100, 0)
```
Comment (jeff-regier) [inline scvi/dataset/loom.py]: This looks like too much "testing code" for the `scvi/` directory. Either 1) we should be loading a testing loom file that has the filtering already done,  or 2) we should subclass LoomDataset in the `test/` directory (can call it `TestLoomDataset`) and overload some method that is a noop in `LoomDataset` to do what's done now in this block.

Comment (jeff-regier) [reply scvi/dataset/loom.py]: Or, is `retina.loom` the whole retina dataset? If so, it's so small that we can probably just use the whole thing for a unit test without any filtering.

Comment (imyiningliu) [reply scvi/dataset/loom.py]: This's not the whole retina dataset. I'll change the test file remove this part of the code.

File: scvi/dataset/retina.py [hunk 4]
```diff
@@ -1,39 +1,39 @@
 import numpy as np
 import scipy.sparse as sp_sparse
+import loompy
 
 from . import GeneExpressionDataset
 
 
 class RetinaDataset(GeneExpressionDataset):
```
Comment (jeff-regier) [inline scvi/dataset/retina.py]: I think we want `RetinaDataset` to inherit from `LoomDataset` now.

File: scvi/dataset/retina.py [hunk 5]
```diff
@@ -1,39 +1,39 @@
...
         second_batch = data_with_info[data_with_info[:, -1] == 1.0]
         first_batch = first_batch[:, :-1]
         second_batch = second_batch[:, :-1]
-        print("Finished preprocessing for Retina %s dataset" % type)
+        print("Finished preprocessing dataset")
         super(RetinaDataset, self).__init__(*GeneExpressionDataset.get_attributes_from_list(
             [sp_sparse.csr_matrix(first_batch[:, :-1]), sp_sparse.csr_matrix(second_batch[:, :-1])],
             list_labels=[first_batch[:, -1], second_batch[:, -1]]))
+
+    def preprocess(self):
```
Comment (jeff-regier) [inline scvi/dataset/retina.py]: I think this code should just appear in `LoomDataset` -- not both places.

File: scvi/dataset/retina.py (hunk 4 above)
Comment (imyiningliu) [reply scvi/dataset/retina.py]: That makes sense. Will do this now 😊

**Comment by imyiningliu** [comment]
I'm not sure what type of data we would want to load for a general `.loom` file so the current LoomDataset class is very specific for `retina.loom`. 

I have another question about implementing `LoomDataset`. Would the users have the datasets locally, or would they also want to pass in an url to the constructor?

File: scvi/dataset/retina.py [hunk 6]
```diff
@@ -20,20 +20,7 @@ def __init__(self, unit_test=False):
         second_batch = data_with_info[data_with_info[:, -1] == 1.0]
         first_batch = first_batch[:, :-1]
         second_batch = second_batch[:, :-1]
-        print("Finished preprocessing dataset")
-        super(RetinaDataset, self).__init__(*GeneExpressionDataset.get_attributes_from_list(
+        print("Finished preprocessing Retina dataset")
+        super(LoomDataset, self).__init__(*GeneExpressionDataset.get_attributes_from_list(
```
Comment (jeff-regier) [inline scvi/dataset/retina.py]: Should be something here like `super(RetinaDataset, self).__init__("data/retina.loom")`

File: scvi/dataset/loom.py [hunk 7]
```diff
@@ -6,12 +6,16 @@
...
+    url = 'https://github.com/YosefLab/scVI-data/raw/master/retina.loom'
 
     def __init__(self, filepath, unit_test=False):
-        self.filepath = filepath if not unit_test else 'tests/data/retina.loom'
+        path, file = os.path.split(filepath)
+        self.save_path = path + '/' if not unit_test else 'tests/data/'
+        self.download_name = file if not unit_test else 'retina.loom'
 
-        data, labels, cell_batches = self.preprocess()
+        data, labels, cell_batches = self.download_and_preprocess()
```
Comment (jeff-regier) [inline scvi/dataset/loom.py]: We just want to call `self.preprocess` in `LoomDataset`...the file should already be downloaded when this point is reached.

File: scvi/dataset/retina.py [hunk 8]
```diff
@@ -1,11 +1,11 @@
...
 
+from .loom import LoomDataset
 from . import GeneExpressionDataset
 
 
-class RetinaDataset(GeneExpressionDataset):
+class RetinaDataset(LoomDataset):
     url = 'https://github.com/YosefLab/scVI-data/raw/master/retina.loom'
 
     def __init__(self, unit_test=False):
```
Comment (jeff-regier) [inline scvi/dataset/retina.py]: Just call `self.download()` in `RetinaDataset`.

**Comment by jeff-regier** [comment]
Yes, users would want to instantiate `LoomDataset` directly with an arbitrary loom file that is stored locally. If you can make `LoomDataset` any more general, please do. But no need to worry too much about it---we can also generalize it once we have some additional loom test files

File: scvi/dataset/retina.py (hunk 8 above)
Comment (imyiningliu) [reply scvi/dataset/retina.py]: Thanks Jeff! The code looks better now ☺️

**Comment by jeff-regier** [comment]
Can we delete retina_batch.npy, retina_counts.npz, and retina_labels.npy from `test/data`?

File: scvi/dataset/retina.py [hunk 9]
```diff
@@ -14,4 +14,4 @@ def __init__(self, unit_test=False):
             self.download()
 
         print("Finished preprocessing Retina dataset")
-        super(RetinaDataset, self).__init__("data/retina.loom")
+        super(RetinaDataset, self).__init__(self.save_path + self.download_name)
```
Comment (jeff-regier) [inline scvi/dataset/retina.py]: Ah, much better :-)

**Comment by jeff-regier** [comment]
Alright, nice work!

**Pull request merged**